### ISSUE TITLE
Exabgp doesn't create a /var/run/exabgp/exabgp.pid in 18.04

### ISSUE THREAD

**Issue by Flygarfisken** [Exa-Networks/exabgp#957]
Hi again,

I am using monit to monitor the processes in ubuntu.

Exabgp doesn't seem to create the exabgp.pid anywhere in the system in version 18.04 and I wonder if you know any workaround?

I have uploaded some information from a working and non working machine

[Non Working Machine.txt](https://github.com/Exa-Networks/exabgp/files/4232918/Non.Working.Machine.txt)
[Working Machine.txt](https://github.com/Exa-Networks/exabgp/files/4232919/Working.Machine.txt)

**Comment by Flygarfisken** [comment]
I can find this in the /etc/init.d/exabgp
PIDFILE=/var/run/$NAME/$NAME.pid

**Comment by thomas-mangin** [comment]
The PID will be either managed by systemd or ExaBGP. I would assume it was caused by a change in the systemd `exabgp.service` file.

`systemctl show -p MainPID exabgp` should provide you with the information.

**Comment by Flygarfisken** [comment]
The PID changes on every reboot and is different on every server - which I can see with the command you provided.
The exabgp 3.4.10 creates a pidfile /var/run/$NAME/$NAME.pid - but the new exabgp 4.0.2 doesn't and I need a pidfile for the software Monit to work.

When I do `cat /etc/init.d/exabgp` in exabgp 4.0.2 I can find this `PIDFILE=/var/run/$NAME/$NAME.pid`
I have installed exabgp by running `apt-get exabgp` and only created an exabgp.conf file

I have uploaded the config of the monit
So basically - the old exabgp creates the pidfile and the new exabgp does not even though I can find the PIDFILE configuration in /init.d/exabgp

[monit.txt](https://github.com/Exa-Networks/exabgp/files/4244288/monit.txt)

**Comment by thomas-mangin** [comment]
@Flygarfisken - I do not package ExaBGP for Ubuntu. I assume the code is the same as for Debian which is kindly maintained by @vincentbernat 

It is possible to modify the systemd service file to have ExaBGP save the PID (or have systemd do it - which is surely the bette path) but it is a sysadmin thing and not a bug with ExaBGP.

I can not support each and every installation of ExaBGP, please look into the `exabgp.daemon.pid` option or on how systemd can save the PID for you (IMHO prefered). 

https://www.freedesktop.org/software/systemd/man/systemd.service.html
```
PIDFile=
Takes a path referring to the PID file of the service. Usage of this option is recommended for services where Type= is set to forking. The path specified typically points to a file below /run/. If a relative path is specified it is hence prefixed with /run/. The service manager will read the PID of the main process of the service from this file after start-up of the service. The service manager will not write to the file configured here, although it will remove the file after the service has shut down if it still exists. The PID file does not need to be owned by a privileged user, but if it is owned by an unprivileged user additional safety restrictions are enforced: the file may not be a symlink to a file owned by a different user (neither directly nor indirectly), and the PID file must refer to a process already belonging to the service.
```

**Comment by Flygarfisken** [comment]
@thomas-mangin Of course, I understand! I appreciate all the help and support I've got so far!

@vincentbernat do you have any suggestions? 

Again, many thanks for all the help!

**Comment by vincentbernat** [comment]
With systemd, ExaBGP is started with `Type=simple` (implicitely). So, no PID file is required, ExaBGP doesn't write one and systemd doesn't write one either. I see two solutions:

1. You teach monit to ask systemd for the main PID (as suggested above)
2. You put in `/etc/systemd/system/exabgp.service.d/override.conf` something like:

```
[Service]
ExecStartPost=sh -c "echo $MAINPID > /var/run/exabgp/exabgp.pid"
```

**Comment by Flygarfisken** [comment]
Hi @vincentbernat and sorry for being so noobish.

Do you mean that I should do like this?`

1. `mkdir /etc/systemd/system/exabgp.service.d` - since the map doesn't exist
2. `nano /etc/systemd/system/exabgp.service.d/override.conf` - since the file doesn't exist
3. Write the service below
 ```
[Service]
ExecStartPost=sh -c "echo $MAINPID > /var/run/exabgp/exabgp.pid"
```
4. save and reboot?



Option 2 (maybe)
I have figured out this too - I found a few files in `/usr/lib/tmpfiles.d/`
I copied one of them and wrote this
```
root@sto-dc01dns02:~# cat /usr/lib/tmpfiles.d/exabgp.conf
d /run/exabgp 0700 root root –
```

The map /var/run/exabgp/ is now created but I don't get any pid-file in it.
Like I said before the PID file changes everytime it reboots the service - is there any way to create a pid-file in this map out of the number (bold-marked) provided below?

```
root@sto-dc01dns02:~# ps ax | grep exabgp
**1070** ?        Ss     0:05 /usr/bin/python3 /usr/sbin/exabgp /etc/exabgp/exabgp.conf
```

**Comment by vincentbernat** [comment]
With first option, use:
```
systemctl daemon-reload
systemctl restart haproxy
```
and you should have a PID in `/var/tun/exabgp/exabgp.pid` which is up-to-date.

**Comment by Flygarfisken** [comment]
@vincentbernat 
Doesn't work... 

root@sto-dc01dns02:~# cat /etc/systemd/system/exabgp.service.d/override.conf
[Service]
ExecStartPost=sh -c "echo $MAINPID > /var/run/exabgp/exabgp.pid"
root@sto-dc01dns02:~# systemctl daemon-reload
root@sto-dc01dns02:~# systemctl restart haproxy
Failed to restart haproxy.service: Unit haproxy.service not found.
root@sto-dc01dns02:~# cd /var/run/exabgp
bash: cd: /var/run/exabgp: No such file or directory

**Comment by vincentbernat** [comment]
Sorry, `systemctl restart exabgp`

**Comment by Flygarfisken** [comment]
@vincentbernat 
Still nothing, tried rebooting the machine too

```
> root@sto-dc01dns02:~# cd /var/run/exabgp
> bash: cd: /var/run/exabgp: No such file or directory
```

**Comment by vincentbernat** [comment]
`systemctl status exabgp` should show you the error. You may change `ExecStartPost=` to:
```
ExecStartPost=sh -c "echo $MAINPID > /var/run/exabgp.pid"
```

**Comment by Flygarfisken** [comment]
The `systemctl status exabgp` is all fine when I do the `systemctl restart exabgp`
Even though I change the `ExecStartPost=` to 
`ExecStartPost=sh -c "echo $MAINPID > /var/run/exabgp.pid"`

[systemstl status exabgp.txt](https://github.com/Exa-Networks/exabgp/files/4248822/systemstl.status.exabgp.txt)

**Comment by thomas-mangin** [comment]
Your guesses are going to be as good as mine. From the output the file was well recognised by systemd.

**Comment by Flygarfisken** [comment]
Hi,
Sorry for late reply, the flue hit me quite bad.
Back to my question:

1. Exabgp version 3.4.10 in Ubuntu 14.04 creates a /var/run/exabgp.pid
2. Exabgp version 4.0.2 in Ubuntu 18.04 doesn't create a pid-file at all because of `Type=simple` - right?
3. I have tried running your commands to create a pid file but it doesn't work

```

> root@server1:~$ ps ax | grep exabgp
>  1095 ?        Ss     0:02 /usr/bin/python3 /usr/sbin/exabgp /etc/exabgp/exabgp.conf
>  1332 ?        S      0:00 /usr/bin/python3 -m exabgp healthcheck --config /etc/exabgp/healthcheck.conf
>  1845 pts/0    S+     0:00 grep --color=auto exabgp

```

I need to figure out how to get the number (in this case) 1095 into a pid file.

This command did not work
`ExecStartPost=sh -c "echo $MAINPID > /var/run/exabgp.pid"`

**Comment by thomas-mangin** [comment]
You could change the exabgp daemonise option and not using systemd simple.

**Comment by Flygarfisken** [comment]
Would you mind explaining?

**Comment by thomas-mangin** [comment]
using `type=forking` and set the env `Environment=exabgp_daemon_daemonize=1`

**Comment by Flygarfisken** [comment]
Where do I find the `type=simple`

I have found `Environment=exabgp_daemon_daemonize` in `/lib/systemd/system/exabgp.service` 
At the moment it says `false` - it should be changed to `1`?

**Comment by vincentbernat** [comment]
Check in `journalctl -u exabgp` why the `ExecStartPost` command may fail. I think putting `/bin/sh` instead of `sh` may fix it.

**Comment by Flygarfisken** [comment]
@vincentbernat 
Many thanks Vicent! Got it working!

1. `mkdir /etc/systemd/system/exabgp.service.d`
2. `nano /etc/systemd/system/exabgp.service.d/override.conf`
3. Paste
```
[Service]
ExecStartPost=/bin/sh -c "echo $MAINPID > /var/run/exabgp.pid"
```
4. Reboot
5. `sudo monit status`

```

> Process 'exabgp'
>   status                       OK
>   monitoring status            Monitored
>   monitoring mode              active
>   on reboot                    start
>   pid                          1135
>   parent pid                   1
>   uid                          113
>   effective uid                113
>   gid                          116
>   uptime                       4m
>   threads                      1
>   children                     1
>   cpu                          1.4%
>   cpu total                    1.4%
>   memory                       0.7% [27.0 MB]
>   memory total                 1.0% [39.2 MB]
>   security attribute           unconfined
>   disk read                    0 B/s [2.4 MB total]
>   disk write                   0 B/s [88 kB total]
>   data collected               Sat, 29 Feb 2020 12:03:28
> 
> Process 'bind'
>   status                       OK
>   monitoring status            Monitored
>   monitoring mode              active
>   on reboot                    start
>   pid                          1083
>   parent pid                   1
>   uid                          112
>   effective uid                112
>   gid                          115
>   uptime                       4m
>   threads                      5
>   children                     0
>   cpu                          0.0%
>   cpu total                    0.0%
>   memory                       0.5% [21.2 MB]
>   memory total                 0.5% [21.2 MB]
>   security attribute           /usr/sbin/named (enforce)
>   disk read                    0 B/s [2.5 MB total]
>   disk write                   0 B/s [16 kB total]
>   port response time           0.389 ms to 185.29.112.1:53 type UDP/IP protocol DNS
>   data collected               Sat, 29 Feb 2020 12:03:28
> 
> System 'sto-dc01dns02'
>   status                       OK
>   monitoring status            Monitored
>   monitoring mode              active
>   on reboot                    start
>   load average                 [0.03] [0.17] [0.09]
>   cpu                          0.8%us 0.2%sy 0.2%wa
>   memory usage                 174.4 MB [4.4%]
>   swap usage                   0 B [0.0%]
>   uptime                       5m
>   boot time                    Sat, 29 Feb 2020 11:59:17
>   data collected               Sat, 29 Feb 2020 12:03:28

```
